src/NuGet.Jobs.GitHubIndexer/GitRepoSearchers/GitHub/GitHubSearcher.cs

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using NuGetGallery;
+using Octokit;
+
+namespace NuGet.Jobs.GitHubIndexer
+{
+    public class GitHubSearcher : IGitRepoSearcher
+    {
+        private readonly ILogger<GitHubSearcher> _logger;
+        private readonly IOptionsSnapshot<GitHubSearcherConfiguration> _configuration;
+        private readonly IGitHubSearchWrapper _searchApiRequester;
+
+        private DateTimeOffset _throttleResetTime;
+
+        public GitHubSearcher(
+            IGitHubSearchWrapper searchApiRequester,
+            ILogger<GitHubSearcher> logger,
+            IOptionsSnapshot<GitHubSearcherConfiguration> configuration)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _searchApiRequester = searchApiRequester ?? throw new ArgumentNullException(nameof(searchApiRequester));
+        }
+
+        private int _minStars => _configuration.Value.MinStars;
+        private int _resultsPerPage => _configuration.Value.ResultsPerPage;
+        private int _maxGithubResultPerQuery => _configuration.Value.MaxGitHubResultsPerQuery;
+
+        private async Task CheckThrottle()
+        {
+            if (_searchApiRequester.GetRemainingRequestCount() == 0)
+            {
+                var sleepTime = _throttleResetTime - DateTimeOffset.Now;
+                _throttleResetTime = DateTimeOffset.Now;
+                if (sleepTime.TotalSeconds > 0)
+                {
+                    _logger.LogInformation($"Waiting {sleepTime.TotalSeconds} seconds to cooldown.");
+                    await Task.Delay(sleepTime);
+                }
+
+                _logger.LogInformation("Resuming search.");
+            }
+        }
+
+        private async Task<SearchRepositoryResult> SearchRepo(SearchRepositoriesRequest request)
+        {
+            _logger.LogInformation("Requesting page {Page} for stars {Stars}", request.Page, request.Stars);
+
+            bool? error = null;
+            GitHubSearchApiResponse response = null;
+            while (!error.HasValue || error.Value)
+            {
+                try
+                {
+                    response = await _searchApiRequester.GetResponse(request);
+                    error = false;
+                }
+                catch (RateLimitExceededException)
+                {
+                    _logger.LogError("Exceeded GitHub RateLimit! Waiting 5 seconds before retrying.");
+                    await Task.Delay(5_000);
```

Comment:
I would use `TimeSpan.FromSeconds(5)` as it's a little more expressive than `5_000`. Our team usually puts these kinds of values as fields on the class. Something like `private static readonly TimeSpan XYZ = TimeSpan.FromSeconds(5);`. Don't forget to update the `LogError` to use this field's value.

```suggestion
                    await Task.Delay(TimeSpan.FromSeconds(5));
```